test/unit/proxy.test.ts

```diff
@@ -102,4 +108,135 @@ describe("proxy", () => {
+  })
+
+  it("should handle invalid routes", async () => {
+    e.post("/wsup", (req, res) => {
+      res.json(req.body)
+    })
+    codeServer = await integration.setup(["--auth=none"], "")
+    const resp = await codeServer.fetch(`${proxyPath}/hello`)
+    expect(resp.status).toBe(404)
+    expect(resp.statusText).toMatch("Not Found")
+  })
+
+  it("should handle errors", async () => {
+    e.use(bodyParser.json({ strict: false }))
+    e.post("/wsup", (req, res) => {
+      throw new Error("BROKEN")
+    })
+    codeServer = await integration.setup(["--auth=none"], "")
+    const resp = await codeServer.fetch(proxyPath, {
+      method: "post",
+      body: JSON.stringify("coder is the best"),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    })
+    expect(resp.status).toBe(500)
+    expect(resp.statusText).toMatch("Internal Server Error")
+  })
+})
+
+describe("proxy (standalone)", () => {
+  const testServer = new httpserver.HttpServer()
+  const proxyServer = proxy
+  let expressServer: express.Express
+
+  beforeAll(async () => {
+    await testServer.listen((req, res) => {
+      proxyServer.web(req, res, { target: `http://localhost:${testServer.port()}` })
+      expressServer(req, res)
+    })
+  })
+
+  afterAll(async () => {
+    await testServer.close()
+    await proxyServer.close()
+  })
+
+  beforeEach(() => {
+    expressServer = express.default()
+  })
+
+  it("should handle errors", async () => {
+    // Notes: I think I'm close
+    // but I think this code is testing that our fake expressServer
+    // is returning a 500
+    // not the proxy
+    expressServer.post("/error", (req, res) => {
+      throw new Error("BROKEN")
+    })
+
+    const resp = await testServer.fetch(`/error`, {
+      method: "post",
+      body: JSON.stringify("coder is the best"),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    })
+    expect(resp.status).toBe(500)
+  })
+})
+
+describe("code-server proxy", () => {
+  const PORT = 9003
+  const PROXY_PORT = 8003
+  const URL = `http://localhost:${PORT}`
+  const PROXY_URL = `http://localhost:${PROXY_PORT}`
+  let testServer: http.Server
+  let proxyTarget: http.Server
+
+  beforeEach(async () => {
+    // Define server and a proxy server
+    testServer = http.createServer((req, res) => {
+      proxy.web(req, res, {
+        target: PROXY_URL,
+      })
+    })
+
+    proxyTarget = http.createServer((req, res) => {
+      res.writeHead(200, { "Content-Type": "text/plain" })
+      res.end()
+    })
+
+    // Start both servers
+    await proxyTarget.listen(PROXY_PORT)
+    await testServer.listen(PORT)
+  })
+
+  afterEach(async () => {
+    await testServer.close()
+    await proxyTarget.close()
+  })
+
+  it("should return a 500 when proxy target errors ", async () => {
+    // Close the proxy target so that proxy errors
+    await proxyTarget.close()
+    const errorResp = await nodeFetch.default(`${URL}/error`)
```

Comment:
As noted during pair programming - Using the `http` and `node-fetch` libraries directly is _easier_ (LoC and maintainability). I have concerns about the custom class using in testing that abstracts these